File: research/experiments/sweep/runs/opt-full-csharp-sonnet45/src/ResultTypes.cs
namespace Optimization;

/// <summary>
/// Shared types and convergence logic for all optimization algorithms.
/// </summary>
public class OptimizeOptions
{
    public double GradTol { get; set; } = 1e-8;
    public double StepTol { get; set; } = 1e-8;
    public double FuncTol { get; set; } = 1e-12;
    public int MaxIterations { get; set; } = 1000;
}

public class OptimizeResult
{
    public double[] X { get; set; } = Array.Empty<double>();
    public double Fun { get; set; }
    public double[]? Gradient { get; set; }
    public int Iterations { get; set; }
    public int FunctionCalls { get; set; }
    public int GradientCalls { get; set; }
    public bool Converged { get; set; }
    public string Message { get; set; } = "";
}

public enum ConvergenceKind
{
    Gradient,
    Step,
    Function,
    MaxIterations,
    LineSearchFailed
}

public class ConvergenceReason
{
    public ConvergenceKind Kind { get; set; }
}

public static class ResultTypes
{
    public static OptimizeOptions DefaultOptions(OptimizeOptions? overrides = null)
    {
        var opts = new OptimizeOptions();
        if (overrides != null)
        {
            opts.GradTol = overrides.GradTol;
            opts.StepTol = overrides.StepTol;
            opts.FuncTol = overrides.FuncTol;
            opts.MaxIterations = overrides.MaxIterations;
        }
        return opts;
    }

    public static OptimizeOptions MergeOptions(OptimizeOptions? overrides)
    {
        if (overrides == null) return new OptimizeOptions();
        return new OptimizeOptions
        {
            GradTol = overrides.GradTol,
            StepTol = overrides.StepTol,
            FuncTol = overrides.FuncTol,
            MaxIterations = overrides.MaxIterations
        };
    }

    public static ConvergenceReason? CheckConvergence(
        double gradNorm, double stepNorm, double funcChange, int iteration, OptimizeOptions opts)
    {
        if (gradNorm <= opts.GradTol)
            return new ConvergenceReason { Kind = ConvergenceKind.Gradient };
        if (stepNorm <= opts.StepTol)
            return new ConvergenceReason { Kind = ConvergenceKind.Step };
        if (funcChange <= opts.FuncTol)
            return new ConvergenceReason { Kind = ConvergenceKind.Function };
        if (iteration >= opts.MaxIterations)
            return new ConvergenceReason { Kind = ConvergenceKind.MaxIterations };
        return null;
    }

    public static bool IsConverged(ConvergenceReason reason)
    {
        return reason.Kind == ConvergenceKind.Gradient ||
               reason.Kind == ConvergenceKind.Step ||
               reason.Kind == ConvergenceKind.Function;
    }

    public static string ConvergenceMessage(ConvergenceReason reason)
    {
        return reason.Kind switch
        {
            ConvergenceKind.Gradient => "Convergence: gradient norm below tolerance",
            ConvergenceKind.Step => "Convergence: step size below tolerance",
            ConvergenceKind.Function => "Convergence: function change below tolerance",
            ConvergenceKind.MaxIterations => "Stopped: reached maximum iterations",
            ConvergenceKind.LineSearchFailed => "Stopped: line search failed",
            _ => "Unknown"
        };
    }
}

File: research/experiments/sweep/runs/opt-medium-python-opus45/src/result_types.py
"""
Shared types and convergence logic used by all optimization algorithms.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class OptimizeOptions:
    grad_tol: float = 1e-8
    step_tol: float = 1e-8
    func_tol: float = 1e-12
    max_iterations: int = 1000


@dataclass
class OptimizeResult:
    x: list[float]
    fun: float
    gradient: Optional[list[float]]
    iterations: int
    function_calls: int
    gradient_calls: int
    converged: bool
    message: str


@dataclass
class ConvergenceReason:
    kind: str  # "gradient", "step", "function", "maxIterations", "lineSearchFailed"


def default_options(**overrides) -> OptimizeOptions:
    """Create OptimizeOptions with defaults, merging any keyword overrides."""
    return OptimizeOptions(**overrides)


def check_convergence(
    grad_norm: float,
    step_norm: float,
    func_change: float,
    iteration: int,
    opts: OptimizeOptions,
) -> Optional[ConvergenceReason]:
    """
    Check convergence criteria in order: gradient -> step -> function -> maxIterations.
    Returns ConvergenceReason or None.
    """
    if grad_norm < opts.grad_tol:
        return ConvergenceReason(kind="gradient")
    if step_norm < opts.step_tol:
        return ConvergenceReason(kind="step")
    if func_change < opts.func_tol:
        return ConvergenceReason(kind="function")
    if iteration >= opts.max_iterations:
        return ConvergenceReason(kind="maxIterations")
    return None


def is_converged(reason: ConvergenceReason) -> bool:
    """True for gradient/step/function; false for maxIterations/lineSearchFailed."""
    return reason.kind in ("gradient", "step", "function")


def convergence_message(reason: ConvergenceReason) -> str:
    """Human-readable convergence message."""
    messages = {
        "gradient": "Converged: gradient norm below tolerance",
        "step": "Converged: step size below tolerance",
        "function": "Converged: function change below tolerance",
        "maxIterations": "Stopped: reached maximum iterations",
        "lineSearchFailed": "Stopped: line search failed to find acceptable step",
    }
    return messages.get(reason.kind, f"Unknown reason: {reason.kind}")
